### ISSUE TITLE
Should get_shortest_paths() return infinite-length paths?

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2222]
**tl;dr** Should infinite-length edges be included when finding shortest paths?

----

Currently `get_shortest_paths()` functions return infinite-length paths. Example in R:

```r
g<-make_graph(~ 1-2-3-1-4-5-6-4)
```

<img width="225" alt="image" src="https://user-images.githubusercontent.com/1212871/193448666-60fada34-980c-4596-9cc9-b502a69619b0.png">

Let's look at paths from 2 to 6 and set the weight the 1-4 bridge to infinity.

```
> shortest_paths(g, 2, 6, weights=c(1,1,1,Inf,1,1,1))$vpath
[[1]]
+ 4/6 vertices, named, from 5a03705:
[1] 2 1 4 6
```

The path is returned despite being infinite-length.

I think this is problematic, and for _most_ use cases it may be more productive to not return the path at all. A common (and very convenient) technique to temporarily remove edges from a graph in a very efficient manner is to set their weights to infinity.

There are other arguments for skipping infinite-weight edges as well. This case is degenerate (leading to multiple "equal-length" paths, all with length infinity). This degeneracy is not handled well. Here are a few examples.

With all-unit weights, the shortest path from 2 to 6 is the following:

```
> shortest_paths(g, 2, 6, weights=c(1,1,1,1,1,1,1))$vpath
[[1]]
+ 4/6 vertices, named, from 5a03705:
[1] 2 1 4 6
```

But if we change the weight of 1-2 to be very large, the path changes as well:

```
> shortest_paths(g, 2, 6, weights=c(10,1,1,1,1,1,1))$vpath
[[1]]
+ 5/6 vertices, named, from 5a03705:
[1] 2 3 1 4 6
```

Setting an infinite weight on the bridge again reverts the shortest path to the original one:

```
> shortest_paths(g, 2, 6, weights=c(10,1,1,Inf,1,1,1))$vpath
[[1]]
+ 4/6 vertices, named, from 5a03705:
[1] 2 1 4 6
```

Of course, it is true that now the weight of both `2 3 1 4 6` and `2 1 4 6` is the same (infinite), but due to allowing infinities, an important invariant is violated in a counterintuitive manner. Namely, if the shortest path from 2 to 6 is `2 1 4 6`, then it must be true that the shortest path from 2 to 1 (an intermediate vertex along the 2-6 path) would be `2 1`. However, this is not true here. The shortest path from 2 to 1 is actually 2 3 1.

Furthermore, if both `2 3 1 4 6` and `2 1 4 6` have the same length, then `get.all.shortest.paths()` should return both. But in the presence of infinity, it does not:

```
> get.all.shortest.paths(g, 2, 6, weights = c(1,1,1,Inf,1,1,1))
$res
$res[[1]]
+ 4/6 vertices, named, from 5a03705:
[1] 2 1 4 6
```

It probably _cannot_ return both because of the above-mentioned broken invariant. (I might be wrong on this, as I haven't studied the algorithm carefully.)

----

For the reasons above, I suggest considering the current behaviour a bug and preventing shortest path algorithms from following infinite-weight edges. When the weight of an edge is infinite, it should be treated as non-existent by shortest path algorithms.

Note: #2221 is related, but is not affected by this issue because the function changed in that PR does not return _paths_, only _distances_.

**Comment by GroteGnoom** [comment]
I don't mind that you will get counterintuitive results when using infinities. They could still be used to simulate something useful I guess.

I do mind a bit that get_all_shortest_paths gives the wrong results. That would require some non-trivial fix. I think this specific problem only pops up when you want 'all' shortest paths. 

Another problem is that `igraph_get_k_shortest_paths` already uses infinities to indicate a removed edge. So that interface is different than others.

So life is easier for us (and possibly some users) if infinite weights means there's no edge. But some users would not be able to simulate a situation the way they want. I'm not sure what to decide, because I feel I have the least knowledge about how people use graphs.

I was thinking maybe users can just use huge floats, but don't you get in trouble there too? I think if you use one float value that's way bigger than others, you'll get almost the same result as with infinity (because huge_float + 1 = huge_float just like inf + 1 = inf). The only difference is that the expected outcome is different.

**Comment by ntamas** [comment]
After reading the examples, I wasn't particularly concerned until I got up to `get.all.shortest.paths()`. For me, an edge with infinite weight is not a nonexistent edge -- it is an edge that costs an infinite amount of resources to traverse, but well, if you do have infinite resources, you can traverse it. :) If you want to exclude an edge from consideration while doing shortest path calculations, you should probably remove it from the graph. (I know it's not efficient). The invariant that you mention simply doesn't hold if there is at least one edge with infinite weight; we could point that out in the documentation and be done with it.

As for the `get.all.shortest.paths()` function, I would consider it a bug, but the easiest way out is to say that this function does not support edges with infinite weights.

---

To make it clear, this is just my opinion, it's not a decision. There are arguments both for and against treating infinities as missing edges, and in my head they pile up with roughly the same weight on both ends of the scale. Forcing me to make a decision _now_ would make me retain the status quo because all other things being equal, this is the one that takes less effort.

Maybe one way out would be to say that `NaN` as a weight means that an edge should be ignored? We could then have the best of both worlds: no changes in the current behaviour w.r.t. infinities, but there would still be a way to indicate that a particular edge should not be used in _any_ shortest paths at all. Unlike for infinity, I see no practical use for `NaN` in a shortest-path context.

**Comment by szhorvat** [comment]
The problem with `get_all_shortest_paths()` and infinite weights is that if the shortest path between two vertices has infinite length, then _all_ paths between those vertices have infinite length. So listing all shortest paths is the same as listing all paths, and the size of the result will suddenly explode.

**Comment by szhorvat** [comment]
The more I look at this, the more I'm convinced that infinite-length edges should be ignored. Doing so still gives an intuitive and easy to understand behaviour.  The behaviour can be documented, and people will easily understand the idea.

Sure, it could also be documented that sub-paths of infinite-length shortest paths are not necessarily shortest, but that is not an easy to understand idea.

There are still ways to find if there is a path from `A` to `B` (i.e. `B` is reachable from `A`).  So it's not a big issue if infinite-length paths are not returned.

We want igraph to be consistent. All shortest-path functions should behave in the same way. That includes functions that compute closeness and betweenness, as well as `get_all_shortest_paths()`. For some of these, it would be quite difficult to implement considering infinite-length paths. It's much easier to exclude infinite-length paths in all functions.

**Comment by szhorvat** [comment]
I'm still of the opinion that the best choice is ignoring infinite-length edges even in functions that return not merely distances but actual paths. Any input on this?

**Comment by ntamas** [comment]
Okay, this is done, but should this go on `master` or `develop` only? Note that the issue is targeted at 0.11, but we can consider this a bugfix and then it can go on `master`.

**Comment by szhorvat** [comment]
I have no objections to making this change on `master`.
